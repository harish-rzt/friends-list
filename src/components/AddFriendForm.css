:local(.addFriendWrapper) {
  display: flex;
  background: #ffffff;
  border-bottom: 1px solid #ABAAAA;
  padding: 4px 20px;
  justify-content: space-between;
}

:local(.AddFriendForm) {
  border: 1px solid #ABAAAA;
  padding: 0 7px;
  margin-left: 12px;
}

:local(.addFriendDropdown) {
  width: 70px;
  padding: 0 5px;
  height: 100%;
}

:local(.addFriendDropdownList) {
  min-width: 100px;
}
