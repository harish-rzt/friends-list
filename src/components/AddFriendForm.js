import classnames from 'classnames';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import styles from './AddFriendForm.css';

const dropDownOptions = [
  { id: 1, value: 'Male' },
  { id: 2, value: 'Female' },
];

class AddFriendForm extends Component {
  render() {
    return (
      <div className={styles.addFriendWrapper}>
        <div className="dropdown">
          <button
            className={classnames('btn btn-default dropdown-toggle', styles.addFriendDropdown)}
            type="button"
            id="dropdownMenu1"
            data-toggle="dropdown"
            aria-haspopup="true"
            aria-expanded="true"
            onClick={this.toggleDropdown.bind(this)}
          >
            {this.state.selectedGender.value}
            <span className="caret" />
          </button>
          {this.state.showDropdownList && <ul
            className={classnames('dropdown-menu show', styles.addFriendDropdownList)}
            aria-labelledby="dropdownMenu1">
            {dropDownOptions && dropDownOptions.map(option =>
              <li key={option.id}>
                <a
                  href='#'
                  className="dropdown-link"
                  onClick={() => {this.handleDropdownChange(option.id);}}
                >
                  {option.value}
                </a>
              </li>,
            )}
          </ul>}
        </div>
        <input
          type="text"
          autoFocus="true"
          className={classnames('form-control', styles.AddFriendForm)}
          placeholder="Type the name of a friend"
          value={this.state.name}
          onChange={this.handleInputChange.bind(this)}
          onKeyDown={this.handleSubmit.bind(this)}
        />
      </div>
    );
  }

  constructor(props, context) {
    super(props, context);
    this.state = {
      name: this.props.name || '',
      showDropdownList: false,
      selectedGender: dropDownOptions[0],
    };
    this.handleDropdownChange = this.handleDropdownChange.bind(this);
  }

  handleInputChange(e) {
    this.setState({ name: e.target.value });
  }

  handleSubmit(e) {
    const name = e.target.value.trim();
    if (e.keyCode === 13 && name.length) {
      const gender = this.state.selectedGender.value;
      this.props.addFriend({ name, gender });
      this.setState({ name: '', selectedGender: dropDownOptions[0] });
    }
  }

  toggleDropdown() {
    this.setState({ showDropdownList: !this.state.showDropdownList });
  }

  handleDropdownChange(id) {
    this.setState({
      selectedGender: dropDownOptions.find(option => option.id === id),
      showDropdownList: false,
    });
  }

}

AddFriendForm.propTypes = {
  addFriend: PropTypes.func.isRequired,
};

export default AddFriendForm;
