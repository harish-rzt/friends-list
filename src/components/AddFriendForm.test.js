import AddFriendForm from './AddFriendForm';

describe('<AddFriendForm />', () => {
  it('Render check with shallow', () => {
    const wrapper = shallow(<AddFriendForm
      addFriend={f => f}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('Render check with render', () => {
    const wrapper = render(<AddFriendForm
      addFriend={f => f}
    />);

    expect(wrapper).toMatchSnapshot();
  });

  it('DOM structure test & default values', () => {
    const wrapper = mount(<AddFriendForm
      addFriend={f => f}
    />);

    // Testing the DOM structure with one input in UI
    const inputsCount = wrapper.find("input").length;
    expect(inputsCount).toEqual(1);

    // Testing the default selected value of input
    const inputValue = wrapper.find("input").text();
    expect(inputValue).toEqual("");

    // Testing the DOM structure with one dropdown in UI
    const dropdownCount = wrapper.find(".dropdown").length;
    expect(dropdownCount).toEqual(1);

    // Testing the default selected value of dropdown
    const defaultDropdownValue = wrapper.find(".dropdown-toggle").text();
    expect(defaultDropdownValue).toEqual("Male");
  });

  it('Simulating dropdown toggle', () => {
    const wrapper = shallow(<AddFriendForm
      addFriend={f => f}
    />);

    // Simulate dropdown toggle
    const dropdownToggle = wrapper.find(".dropdown-toggle");
    expect(wrapper.find('.dropdown-menu').length).toEqual(0);
    dropdownToggle.simulate("click");
    expect(wrapper.find('.dropdown-menu').length).toEqual(1);
  });

  it('Simulating dropdown value change', () => {
    const wrapper = shallow(<AddFriendForm
      addFriend={f => f}
    />);

    // Simulate dropdown toggle
    const dropdownToggle = wrapper.find(".dropdown-toggle");

    // before clicking on dropdown
    expect(wrapper.find('.dropdown-menu').length).toEqual(0);

    // clicking on dropdown
    dropdownToggle.simulate("click");

    // checking for dropdown options count
    const dropdownLinks = wrapper.find('.dropdown-link');
    expect(dropdownLinks.length).toEqual(2);

    // after clicking on dropdown the dropdown menu will append to DOM
    expect(wrapper.find('.dropdown-menu').length).toEqual(1);

    // checking for default dropdown selected value
    const defaultDropdownValue = wrapper.find(".dropdown-toggle").text();
    expect(defaultDropdownValue).toEqual("Male");
    expect(wrapper.state().selectedGender.value).toEqual("Male");

    // checking for default dropdown selected value
    dropdownLinks.at(1).simulate('click');

    // after selecting Female option in the dropdown
    const updatedDropdownValue = wrapper.find(".dropdown-toggle").text();
    expect(updatedDropdownValue).toEqual("Female");
    expect(wrapper.state().selectedGender.value).toEqual("Female");
  });

  it('Simulating input value change', () => {
    const wrapper = shallow(<AddFriendForm
      addFriend={f => f}
    />);
    expect(wrapper.state().name).toEqual("");
    wrapper.find("input").simulate('change', {target: {value: 'test'}});
    expect(wrapper.state().name).toEqual("test");
  });

  it('Simulating input on submit', () => {
    const spy = sinon.spy();
    const wrapper = mount(<AddFriendForm
      addFriend={spy}
    />);
    const input = wrapper.find('input');
    expect(spy.callCount).toBe(0);
    input.simulate('change', { target: { value: 'abcdefg'} });
    input.simulate('keydown', { keyCode: 13 });
    expect(spy.callCount).toBe(1);
  });
});