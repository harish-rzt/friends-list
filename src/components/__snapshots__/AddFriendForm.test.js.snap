exports[`<AddFriendForm /> Render check with render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "dropdown",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "aria-expanded": "true",
              "aria-haspopup": "true",
              "class": "btn btn-default dropdown-toggle",
              "data-toggle": "dropdown",
              "id": "dropdownMenu1",
              "type": "button",
            },
            "children": Array [
              Object {
                "data": "Male",
                "next": Object {
                  "attribs": Object {
                    "class": "caret",
                  },
                  "children": Array [],
                  "name": "span",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
              Object {
                "attribs": Object {
                  "class": "caret",
                },
                "children": Array [],
                "name": "span",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": Object {
                  "data": "Male",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
            ],
            "name": "button",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "aria-expanded": undefined,
              "aria-haspopup": undefined,
              "class": undefined,
              "data-toggle": undefined,
              "id": undefined,
              "type": undefined,
            },
            "x-attribsPrefix": Object {
              "aria-expanded": undefined,
              "aria-haspopup": undefined,
              "class": undefined,
              "data-toggle": undefined,
              "id": undefined,
              "type": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "form-control",
            "placeholder": "Type the name of a friend",
            "type": "text",
            "value": "",
          },
          "children": Array [],
          "name": "input",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "placeholder": undefined,
            "type": undefined,
            "value": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "form-control",
          "placeholder": "Type the name of a friend",
          "type": "text",
          "value": "",
        },
        "children": Array [],
        "name": "input",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "dropdown",
          },
          "children": Array [
            Object {
              "attribs": Object {
                "aria-expanded": "true",
                "aria-haspopup": "true",
                "class": "btn btn-default dropdown-toggle",
                "data-toggle": "dropdown",
                "id": "dropdownMenu1",
                "type": "button",
              },
              "children": Array [
                Object {
                  "data": "Male",
                  "next": Object {
                    "attribs": Object {
                      "class": "caret",
                    },
                    "children": Array [],
                    "name": "span",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
                Object {
                  "attribs": Object {
                    "class": "caret",
                  },
                  "children": Array [],
                  "name": "span",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": Object {
                    "data": "Male",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
              ],
              "name": "button",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "aria-expanded": undefined,
                "aria-haspopup": undefined,
                "class": undefined,
                "data-toggle": undefined,
                "id": undefined,
                "type": undefined,
              },
              "x-attribsPrefix": Object {
                "aria-expanded": undefined,
                "aria-haspopup": undefined,
                "class": undefined,
                "data-toggle": undefined,
                "id": undefined,
                "type": undefined,
              },
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "placeholder": undefined,
          "type": undefined,
          "value": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "placeholder": undefined,
          "type": undefined,
          "value": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`<AddFriendForm /> Render check with shallow 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AddFriendForm
    addFriend={[Function]} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="dropdown">
          <button
                    aria-expanded="true"
                    aria-haspopup="true"
                    className="btn btn-default dropdown-toggle"
                    data-toggle="dropdown"
                    id="dropdownMenu1"
                    onClick={[Function]}
                    type="button">
                    Male
                    <span
                              className="caret" />
          </button>
</div>,
        <input
          autoFocus="true"
          className="form-control"
          onChange={[Function]}
          onKeyDown={[Function]}
          placeholder="Type the name of a friend"
          type="text"
          value="" />,
      ],
      "className": undefined,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              aria-expanded="true"
              aria-haspopup="true"
              className="btn btn-default dropdown-toggle"
              data-toggle="dropdown"
              id="dropdownMenu1"
              onClick={[Function]}
              type="button">
              Male
              <span
                            className="caret" />
</button>,
            false,
          ],
          "className": "dropdown",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "aria-expanded": "true",
              "aria-haspopup": "true",
              "children": Array [
                "Male",
                <span
                  className="caret" />,
              ],
              "className": "btn btn-default dropdown-toggle",
              "data-toggle": "dropdown",
              "id": "dropdownMenu1",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": Array [
              "Male",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "caret",
                },
                "ref": null,
                "rendered": null,
                "type": "span",
              },
            ],
            "type": "button",
          },
          false,
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "autoFocus": "true",
          "className": "form-control",
          "onChange": [Function],
          "onKeyDown": [Function],
          "placeholder": "Type the name of a friend",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="dropdown">
            <button
                        aria-expanded="true"
                        aria-haspopup="true"
                        className="btn btn-default dropdown-toggle"
                        data-toggle="dropdown"
                        id="dropdownMenu1"
                        onClick={[Function]}
                        type="button">
                        Male
                        <span
                                    className="caret" />
            </button>
</div>,
          <input
            autoFocus="true"
            className="form-control"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Type the name of a friend"
            type="text"
            value="" />,
        ],
        "className": undefined,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                aria-expanded="true"
                aria-haspopup="true"
                className="btn btn-default dropdown-toggle"
                data-toggle="dropdown"
                id="dropdownMenu1"
                onClick={[Function]}
                type="button">
                Male
                <span
                                className="caret" />
</button>,
              false,
            ],
            "className": "dropdown",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "aria-expanded": "true",
                "aria-haspopup": "true",
                "children": Array [
                  "Male",
                  <span
                    className="caret" />,
                ],
                "className": "btn btn-default dropdown-toggle",
                "data-toggle": "dropdown",
                "id": "dropdownMenu1",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": Array [
                "Male",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "className": "caret",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "span",
                },
              ],
              "type": "button",
            },
            false,
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoFocus": "true",
            "className": "form-control",
            "onChange": [Function],
            "onKeyDown": [Function],
            "placeholder": "Type the name of a friend",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "prevContext": true,
          },
          "getChildContext": Object {
            "calledByRenderer": true,
          },
        },
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
